gateway/api/views/files.py

```diff
@@ -67,38 +69,196 @@ def check_user_has_provider(self, user, provider_name):
+            return None
+
+        function = Program.objects.filter(
+            title=function_title, provider=provider
+        ).first()
+        if function is None:
+            logger.error(
+                "Function [%s/%s] does not exist.", provider_name, function_title
+            )
+            return None
+        return function
+
+    def user_has_access_to_provider_function(self, user, function: Program) -> bool:
+        """
+        This method returns True or False if the user has access to the provider function.
+
+        Args:
+            user: the user that is doing the request
+            funtion (Program): the Qiskit Function that is going to be checked
+
+        Returns:
+            bool: boolean value that verifies if the user has access or not to the function
+        """
+
+        instances = function.instances.all()
+        user_groups = user.groups.all()
+        has_access = any(group in instances for group in user_groups)
+        if not has_access:
+            logger.error(
+                "User [%s] has no access to function [%s/%s].",
+                user.id,
+                function.provider.name,
+                function.title,
+            )
+        return has_access
+
     def list(self, request):
-        """List of available for user files."""
-        response = Response(
-            {"message": "Requested file was not found."},
-            status=status.HTTP_404_NOT_FOUND,
-        )
-        files = []
+        """
+        It returns a list with the names of available files for the user directory:
+            it will look under its username or username/provider_name/function_title
+        """
+
         tracer = trace.get_tracer("gateway.tracer")
         ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
         with tracer.start_as_current_span("gateway.files.list", context=ctx):
-            user_dir = request.user.username
-            provider_name = request.query_params.get("provider")
-            if provider_name is not None:
-                if self.check_user_has_provider(request.user, provider_name):
-                    user_dir = provider_name
+            username = request.user.username
+            provider_name = sanitize_name(request.query_params.get("provider"))
+            function_title = sanitize_name(request.query_params.get("function"))
+            working_dir = USER_STORAGE
+
+            function = None
+            if provider_name:
+                function = self.get_provider_function(
+                    provider_name=provider_name, function_title=function_title
+                )
+                if function:
+                    if not self.user_has_access_to_provider_function(
+                        request.user, function
+                    ):
+                        return Response(
+                            {
+                                "message": "You don't have access to this Qiskit Function."
```

Comment:
Should we return always `404 - doesn't exist` always for security? Usually, if some resource is protected, you don't want let others know that it exist.

Reply:
Solved in [a0210c1](https://github.com/Qiskit/qiskit-serverless/pull/1529/commits/a0210c11559762adcb917869fb9fc3dfb5d8b47a)